allow passing extra measurement functions to regionprops and regionprops_table

VolkerH opened the pull request:
## Description

This is a WIP implements the feature request discussed here: https://github.com/scikit-image/scikit-image/issues/4466 

It extends `regionprops` and `regionprops_table` by adding an extra argument, which is an iterable of callables/functions.

A user might want to do this if a measurment they require is not included in the built-in properties. By specifying these functions users won't have to duplicate the functionality that loops over the regions. Possible use cases: 

* provide a measurement function that is too specifically focused to include in the skimage code base
* provide measurement function that is under an incompatible license
* provide a faster implementation for an existing measurement function

Each provided function must either take one (`region`) or two (`region`, `intensity_image`) (names for illustration, this is by order) arguments. 

In the issue, there was some discussion on whether the name and return dtype of the measurement function need to be passed in addition, as these are needed to build the regionprops table. Both the name of new property and the dtype are now inferred.
The name is inferred simply as the function name, and a warning is issued if there is a name clash. The return dtype is determined by passing a tiny test image to the function and inspecting the return value.

At this stage, the functionality is working but tests and examples still need to be added.

Code contributions:
Much of code in this PR was co-authored by @jni during a VS Code LiveShare session (somehow VS code only appended the co-authorship note to one of the commit messages).

## Checklist


- DONE  [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- TODO Gallery example in `./doc/examples` (new features only)
- SKIP Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- TODO Unit tests
- DONE Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)


## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

pep8speaks commented:
Hello @VolkerH! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/measure/_regionprops.py`](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py):

> [Line 141:1](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L141): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 167:16](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L167): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 174:53](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L174): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 683:76](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L683): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 685:73](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L685): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 687:76](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L687): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 809:76](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L809): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 811:73](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L811): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 813:76](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L813): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 815:1](https://github.com/scikit-image/scikit-image/blob/84a51f2e61544f19801cc92e1129aac0b162963e/skimage/measure/_regionprops.py#L815): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

jni commented:
@scikit-image/core before we keep moving on this, it would be great to get some consensus on the API. Currently, it looks like this:

```python
rp = regionprops(
    label_image,
    intensity_image,
    extra_properties=(mesh_surface, porosity, average_surface_curvature),
)
```

Those properties can then be accessed with `rp[0].mesh_surface` etc. I think it's pretty nice!

stefanv commented:
> @scikit-image/core before we keep moving on this, it would be great to get some consensus on the API. Currently, it looks like this:
> 
> ```python
> rp = regionprops(
>     label_image,
>     intensity_image,
>     extra_properties=(mesh_surface, porosity, average_surface_curvature),
> )
> ```
> 
> Those properties can then be accessed with `rp[0].mesh_surface` etc. I think it's pretty nice!

Shouldn't that be a dict instead of a tuple?

jni commented:
@stefanv 

> Shouldn't that be a dict instead of a tuple?

dict mapping what to what? It's a sequence of functions. We *can* make a dict mapping name -> function, but Python makes that unnecessary/redundant since we can use `func.__name__`.

stefanv commented:
I guess I was thinking about lambdas, but then it's easy enough to just define a new function.

jni commented:
> I guess I was thinking about lambdas

Yes, I would leave this concern for future PRs. We could also accept *both* a sequence and a dict, but I fear that complicates things.

VolkerH commented:
added

* tests
* docstring examples

also 
* managed to shut up pep8speaks (for the moment)
* corrected one serious bug that only showed up when testing with an intensity image and at least two different labels. As the SAMPLE arrays in the tests only have a single label that was not picked up. Added a SAMPLE_MULTIPLE array in the tests.

I'm not sure whether a gallery example makes sense as there is no image to show. 

@jni, do you want to have another look before I remove the WIP and flag this as ready for review?

VolkerH commented:
Anyway, I think this is ready for review.

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -166,6 +228,45 @@ def __init__(self, slice, label, label_image, intensity_image,
+            elif n_args == 1:
+                return func(self.image)
+            else:
+                raise AttributeError(f"Function provided for custom regionprop"
+                                     " {attr} takes incorrect number of"
+                                     " arguments."
+                                     )
```
```suggestion
                raise AttributeError(
                    f"Function provided for custom regionprop {attr} takes {n_args} "
                    "arguments, but only 1 or 2 are supported."
                )
```

jni replied:
(The f-ness of a string does not propagate across multiple lines, fyi.)

VolkerH replied:
I did not know that ! Interesting.

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -638,33 +755,62 @@ def regionprops_table(label_image, intensity_image=None,
 
+    If we want to measure a feature that does not come as a built-in
+    property, we can define custom functions and pass them as
+    ``extra_properies``. Let's create a custom function that measures
+    the inensity quartiles in a region:
```
```suggestion
    ``extra_properies``. For example, we can create a custom function
    that measures the intensity quartiles in a region:
```

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -638,33 +755,62 @@ def regionprops_table(label_image, intensity_image=None,
+
+    >>> from skimage import data, util, measure
+    >>> import numpy as np
+    >>> def quartiles(regionmask, intensity):
+    >>>     return np.percentile(intensity[regionmask], q=(25, 50, 75))
```
```suggestion
    >>> def quartiles(regionmask, intensity):
    ...     return np.percentile(intensity[regionmask], q=(25, 50, 75))
```

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -638,33 +755,62 @@ def regionprops_table(label_image, intensity_image=None,
+    >>>
+    >>> image = data.coins()
+    >>> label_image = measure.label(image > 110, connectivity=image.ndim)
+    >>> props = measure.regionprops_table(label_image, intensity_image=image,
+    >>>                                   properties=('label',),
+    >>>                                   extra_properties=(quartiles,))
```
```suggestion
    >>> props = measure.regionprops_table(label_image, intensity_image=image,
    ...                                   properties=('label',),
    ...                                   extra_properties=(quartiles,))
```

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -638,33 +755,62 @@ def regionprops_table(label_image, intensity_image=None,
+    >>>                                   extra_properties=(quartiles,))
+    >>> import pandas as pd # doctest: +SKIP
+    >>> pd.DataFrame(props).head() # doctest: +SKIP
+       label  quartiles-0  quartiles-1  quartiles-2
+0      1       117.00        123.0        130.0
+1      2       111.25        112.0        114.0
+2      3       111.00        111.0        111.0
+3      4       111.00        111.5        112.5
+4      5       112.50        113.0        114.0
```
```suggestion
           label  quartiles-0  quartiles-1  quartiles-2
    0      1       117.00        123.0        130.0
    1      2       111.25        112.0        114.0
    2      3       111.00        111.0        111.0
    3      4       111.00        111.5        112.5
    4      5       112.50        113.0        114.0
```

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -872,6 +1024,20 @@ def regionprops(label_image, intensity_image=None, cache=True,
+    >>> import numpy as np
+    >>> img = util.img_as_ubyte(data.coins()) > 110
+    >>> label_img = label(img, connectivity=img.ndim)
+    >>> def pixelcount(regionmask):
+    >>>     return np.sum(regionmask)
```
```suggestion
    >>> def pixelcount(regionmask):
    ...     return np.sum(regionmask)
```

jni commented on skimage/measure/_regionprops.py:
```diff
@@ -872,6 +1024,20 @@ def regionprops(label_image, intensity_image=None, cache=True,
     >>> props[0]['centroid']
     (22.72987986048314, 81.91228523446583)
 
+    Add custom measurements by passing functions as ``extra_properties``
```
```suggestion
    Add custom measurements by passing functions as ``extra_properties``:

```

VolkerH commented:
Thanks for those changes.  Writing code in comments presents different challenges from just writing code.

jni commented:
Everything is passing! @scikit-image/core this is ready for review!

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -166,6 +228,45 @@ def __init__(self, slice, label, label_image, intensity_image,
+                return func(self.image)
+            else:
+                raise AttributeError(
+                    f"Function provided for custom regionprop {attr} takes {n_args} "
+                    "arguments, but only 1 or 2 are supported."
```
```suggestion
                    "Custom regionprop function's number of arguments must be 1 or 2"
                    f"but {attr} takes {n_args} arguments."
```

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -693,7 +839,13 @@ def regionprops(label_image, intensity_image=None, cache=True,
+        the dtype is inferred by calling the function on a small sample.
+        If the name of an extra property clashes with the name of an existing
+        property the extra property wil not be visible and a UserWarning is
+        issued.
```
Shouldn't we specify that the first argument of the property function must be the mask of the considered region and the optional second argument must be the corresponding image?

VolkerH replied:
yes, we should. So far that could only be derived from the examples. I just added this, happy for suggestions about how to phrase it better.

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -166,6 +228,45 @@ def __init__(self, slice, label, label_image, intensity_image,
+                    f"Extra property '{name}' is shadowed by existing "
+                    "property and will be inaccessible. Consider renaming it."
+                )
+                warn(msg)
```
Can't we dynamically edit `COL_DTYPES` to avoid the handling of  dtype for extra properties in `_props_to_dict`?
```suggestion
                warn(msg)
            else:
                COL_DTYPES[func.__name__] = _infer_regionprop_dtype(
                    func,
                    intensity=intensity_image is not None,
                    ndim=self._ndim,
                )
```

jni replied:
@rfezzani  here we chose to preserve existing behaviour rather than try to improve it, which could have unintended consequences. I suggest leaving dynamic inference of existing properties for a future PR.

VolkerH replied:
@jni was faster than me. In general, I don't like modifying a global (in the namespace) state. That state change will persist in `COL_DTYPES` potentially leading to the "unintended consequences" that @jni mentioned.

rfezzani replied:
The suggestion is not an improvement to existing behavior but the modification of the extra properties (new feature) dtype handeling ^^. Avoiding namespace variable modification is a good point, no problem ;)

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -579,6 +689,13 @@ def regionprops_table(label_image, intensity_image=None,
+        the dtype is inferred by calling the function on a small sample.
+        If the name of an extra property clashes with the name of an existing
+        property the extra property wil not be visible and a UserWarning is
+        issued.
```
Same [question](https://github.com/scikit-image/scikit-image/pull/4810/files#r449624730)?

VolkerH commented:
Thanks very much for the review and the suggestions @rfezzani, I made the suggested changes apart from the dynamic addition to `COL_DTYPES` (see comments there).

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -872,6 +1028,20 @@ def regionprops(label_image, intensity_image=None, cache=True,
+    >>> img = util.img_as_ubyte(data.coins()) > 110
+    >>> label_img = label(img, connectivity=img.ndim)
+    >>> def pixelcount(regionmask):
+    >>>     return np.sum(regionmask)
```
This is causing the CI failure
```suggestion
            return np.sum(regionmask)
```

VolkerH replied:
Thanks a lot for finding this. I didn't look at the failing Travis build log as it has been failing with matplotlib for a while and assumed that was still the case. Will commit your suggestion.
I'm curious: are the other CI builds not building documentation or why are they all give the green tick?

rfezzani replied:
>  I'm curious: are the other CI builds not building documentation or why are they all give the green tick?

They are, and they crashed now :rofl:

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -872,6 +1028,20 @@ def regionprops(label_image, intensity_image=None, cache=True,
+    >>> img = util.img_as_ubyte(data.coins()) > 110
+    >>> label_img = label(img, connectivity=img.ndim)
+    >>> def pixelcount(regionmask):
+            return np.sum(regionmask)
```
```suggestion
...         return np.sum(regionmask)
```
Now it should be OK ;)

rfezzani commented on skimage/measure/_regionprops.py:
```diff
@@ -872,6 +1028,20 @@ def regionprops(label_image, intensity_image=None, cache=True,
+    >>> img = util.img_as_ubyte(data.coins()) > 110
+    >>> label_img = label(img, connectivity=img.ndim)
+    >>> def pixelcount(regionmask):
+...         return np.sum(regionmask)
```
```suggestion
    ...     return np.sum(regionmask)
```

rfezzani commented:
:tada: The CI is green again (almost :p). Merging!
Thank you again @VolkerH ^^

VolkerH commented:
Thank you for fixing the Sphinx build and the review @rfezzani . Credit should also go to @jni !

refack commented:
Hello @VolkerH,

Firstly I would like to say that I love this feature, thank you for this 🙏.
Secondly I'd like to ask if there's a reason why you chose not to pass the `RegionProperties` object to the functions? (a.k.a. `self` in the context of their evaluation)
https://github.com/scikit-image/scikit-image/blob/3a4dc584d062347d9ab6bcd0910aeb7b40bd9623/skimage/measure/_regionprops.py#L253


Just for background, for my use case I need to know the `label` of the region in order to cross corelate it.

VolkerH commented:
Hi @refack ,

glad you like it, appreciate the feedback that it is useful.

As for your second question: it simply did not occur to me at the time and I wanted a simple function signature for the `extra_properties` function. Someone who implements such a function, should not need to know about any other input other than the mask and potentially the intensity image for the region.

If your extra property depends on other properties calculated for the region, I am not sure whether passing `self` will be so good as I don't think it can be guarantueed that those other properties will already have been calculated. Also, passing the object might allow an incorrectly implemented `extra_properties` function to modify the object (and thus other properties). 

If your desired result depends on several properties, is it possible to calculate if from several columns of the `regionsprops_table`?

I have not looked at the regionprops code since last July (and that is the only time I looked at it) so I might be missing something. It may also be possible to add what you want without breaking things. But @jni (who deserves at least equal credit for implementing this feature) will be better positioned to answer that.

jni commented:
> it simply did not occur to me at the time

Me neither! But just today I found that I could have used it!

> I don't think it can be guaranteed that those other properties will already have been calculated.

Actually this is not a concern — all the properties are dynamically calculated. It would be a concern if you wanted to have chains of custom functions that depended on each other. Then you would have to make sure that you provided them in the order they are needed.

@refack I think this is certainly worth considering. Would you mind creating an issue specifically for this (e.g. "Proposal: optionally pass Regionprops object to custom properties") so that we can discuss the best way to do this? imho the argument should not be called "self" as that is a reserved-by-convention term in Python. Perhaps simply `props`?
